### PR TITLE
Fixed numerical problems with bubble/dew point pressures

### PR DESCRIPTION
**Pull request by babrodtk**
This addresses the issues with the dew / bubble point pressures causing excessive terminal output.

The issues were multiple:

 * The tables `RsTable` and `RvTables` may have flat regions in them, thus with a zero derivative. Previously it threw an exception + log problem, now it accepts the initial guess/evaluation as the value.

 * The `saturationPressure_` tables may have multiple identical `x` (Rs/Rv) values, which essentially makes the `Tabulated1Dfunction` return `nan`/`inf` when trying to evaluate. These duplicate values are now pruned when creating the table.

I have not observed any side effects on the tests on my side.

### CODE REVIEW

**Comment by atgeirr** [comment]
jenkins build this with downstreams please

File: opm/material/common/Tabulated1DFunction.hpp [hunk 1]
```diff
@@ -272,7 +272,12 @@ class Tabulated1DFunction
         Scalar y0 = yValues_[segIdx];
         Scalar y1 = yValues_[segIdx + 1];
 
-        return y0 + (y1 - y0)*(x - x0)/(x1 - x0);
+        // Corner case where table is poorly defined
+        // with multiple y values for a single x value
+        if (x0 == x1)
```
Comment (atgeirr) [inline opm/material/common/Tabulated1DFunction.hpp]: I'd prefer braces, but I don't think @andlaus requires them...

File: opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp [hunk 2]
```diff
@@ -529,16 +529,13 @@ class LiveOilPvt
...
-
-            if (!Opm::isfinite(delta)) {
-                 std::stringstream errlog;
-                 errlog << "Obtain delta with non-finite value for Rs = " << Rs << " in the "
-                        << i << "th iteration during finding saturation pressure iteratively";
-                 OpmLog::problem("wetgas NaN delta value", errlog.str());
-                 OPM_THROW_NOLOG(NumericalProblem, errlog.str());
+            //If the derivative is zero Newton will not converge, so simply return our best guess instead
+            if (Toolbox::scalarValue(fPrime) == 0.0) {
+                return pSat;
```
Comment (atgeirr) [inline opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp]: I think perhaps a "better" value to return would be the lowest pressure for which the derivative is zero. It should not matter much, though.

File: opm/material/fluidsystems/blackoilpvt/WetGasPvt.hpp [hunk 3]
```diff
@@ -557,16 +557,13 @@ class WetGasPvt
...
-
-            if (!Opm::isfinite(delta)) {
-                 std::stringstream errlog;
-                 errlog << "Obtain delta with non-finite value for Rv = " << Rv << " in the "
-                        << i << "th iteration during finding saturation pressure iteratively";
-                 OpmLog::problem("wetgas NaN delta value", errlog.str());
-                 OPM_THROW_NOLOG(NumericalProblem, errlog.str());
+            //If the derivative is zero Newton will not converge, so simply return our best guess instead
+            if (Toolbox::scalarValue(fPrime) == 0.0) {
+                return pSat;
```
Comment (atgeirr) [inline opm/material/fluidsystems/blackoilpvt/WetGasPvt.hpp]: Same applies here: could be better to report lowest possible pressure.

File: opm/material/fluidsystems/blackoilpvt/WetGasPvt.hpp [hunk 4]
```diff
@@ -605,9 +605,15 @@ class WetGasPvt
...
 
-            std::pair<Scalar, Scalar> val(Rv, pSat);
+            Pair val(Rv, pSat);
             pSatSamplePoints.push_back(val);
         }
+
+        //Prune duplicate Rv values (can occur, and will cause problems in further interpolation)
+        auto x_coord_comparator = [](const Pair& a, const Pair& b) { return a.first == b.first; };
+        auto last = std::unique(pSatSamplePoints.begin(), pSatSamplePoints.end(), x_coord_comparator);
+        pSatSamplePoints.erase(last, pSatSamplePoints.end());
```
Comment (atgeirr) [inline opm/material/fluidsystems/blackoilpvt/WetGasPvt.hpp]: This code (and the live oil version) should be unit tested, perhaps as simply as adding extra (redundant) data to an existing test?

File: opm/material/common/Tabulated1DFunction.hpp [hunk 5]
```diff
@@ -272,7 +272,12 @@ class Tabulated1DFunction
...
+            return y0;
```
Comment (atgeirr) [inline opm/material/common/Tabulated1DFunction.hpp]: Is this change really necessary, when the tables are pruned?

Comment (babrodtk) [inline opm/material/common/Tabulated1DFunction.hpp]: I don't think it will be needed, no.

File: opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp (hunk 2 above)
Comment (babrodtk) [inline opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp]: My initial tests with this were not very successful. I'm running a simulation now to see if it is actually better or not.

File: opm/material/fluidsystems/blackoilpvt/WetGasPvt.hpp (hunk 4 above)
Comment (babrodtk) [inline opm/material/fluidsystems/blackoilpvt/WetGasPvt.hpp]: I agree, though I am a bit uncertain as to how to make that test. I've looked at test_eclblackoilpvt.cpp but it's not clear to me how to actually design such a test.

**Comment by babrodtk** [comment]
I've checked the results of this PR against a large case (with PBUB/PDEW enabled) run with master. The convergence, overall well iterations, linearizations, newton iterations, and linear iterations are identical. The solver time difference is one second after one wall clock hour of simulation, so I have not seen any performance impact either.

File: opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp (hunk 2 above)
Comment (babrodtk) [inline opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp]: After some testing, it appears to me that the results returning the minimum pressure, using the following code, 
```
            // If the derivative is zero Newton will not converge,
            // so simply return the smallest value.
            if (Toolbox::scalarValue(fPrime) == 0.0) {
                //Find interval where derivative is zero
                const auto& table = saturationPressure_[regionIdx];
                size_t index = RvTable.findSegmentIndex(Toolbox::scalarValue(pSat));
                Scalar x0 = RvTable.xAt(index);
                Scalar x1 = RvTable.xAt(index+1);

                //Evaluate at end points and chose minimum
                const Evaluation y0 = table.eval(x0, /*extrapolate=*/true);
                const Evaluation y1 = table.eval(x1, /*extrapolate=*/true);
                if (std::abs(Toolbox::scalarValue(y0)) < std::abs(Toolbox::scalarValue(y1))) {
                    return y0;
                }
                else {
                    return y1;
                }
            }
```
gives identical results to just returning pSat. This, however, requires that the previously private function Tabulated1DFunction::findSegmentIndex(value) is made public. I'm open for suggestions on what is best here.

File: opm/material/common/Tabulated1DFunction.hpp (hunk 1 above)
Comment (andlaus) [inline opm/material/common/Tabulated1DFunction.hpp]: right: I don't care, so I'll just follow the judgement of the patch's author. I'm less lenient about identation, though.(which is fine here, BTW)

File: opm/material/common/Tabulated1DFunction.hpp (hunk 5 above)
Comment (andlaus) [inline opm/material/common/Tabulated1DFunction.hpp]: Because it is simpler I prefer to have it removed in this case.

File: opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp [hunk 6]
```diff
@@ -529,16 +529,13 @@ class LiveOilPvt
...
-            const Evaluation& delta = f/fPrime;
-
-            if (!Opm::isfinite(delta)) {
-                 std::stringstream errlog;
-                 errlog << "Obtain delta with non-finite value for Rs = " << Rs << " in the "
-                        << i << "th iteration during finding saturation pressure iteratively";
-                 OpmLog::problem("wetgas NaN delta value", errlog.str());
-                 OPM_THROW_NOLOG(NumericalProblem, errlog.str());
+            //If the derivative is zero Newton will not converge, so simply return our best guess instead
+            if (Toolbox::scalarValue(fPrime) == 0.0) {
```
Comment (andlaus) [inline opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp]: I suppose you should do this using

```c++
if (std::abs(Toolbox::scalarValue(fPrime)) < 1e-30)
```

otherwise GCC spits out some warnings?

File: opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp [hunk 7]
```diff
@@ -579,9 +579,15 @@ class LiveOilPvt
...
 
-            std::pair<Scalar, Scalar> val(Rs, pSat);
+            Pair val(Rs, pSat);
             pSatSamplePoints.push_back(val);
         }
+
+        //Prune duplicate Rs values (can occur, and will cause problems in further interpolation)
+        auto x_coord_comparator = [](const Pair& a, const Pair& b) { return a.first == b.first; };
+        auto last = std::unique(pSatSamplePoints.begin(), pSatSamplePoints.end(), x_coord_comparator);
+        pSatSamplePoints.erase(last, pSatSamplePoints.end());
```
Comment (andlaus) [inline opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp]: we should think about whether this should be done by `Opm::Tabulated1DFunction` and `Opm::Spline`...

File: opm/material/fluidsystems/blackoilpvt/WetGasPvt.hpp (hunk 3 above)
Comment (andlaus) [inline opm/material/fluidsystems/blackoilpvt/WetGasPvt.hpp]: doesn't GCC produce warnings here?

File: opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp (hunk 7 above)
Comment (babrodtk) [inline opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp]: I don't really understand what you mean here. Do you mean that both Tabulated1DFunction and Spline should do this, or that perhaps it should be done somewhere else?

File: opm/material/fluidsystems/blackoilpvt/WetGasPvt.hpp (hunk 3 above)
Comment (babrodtk) [inline opm/material/fluidsystems/blackoilpvt/WetGasPvt.hpp]: I didn't notice a warning with this one, but I can add the fix you suggest.

Comment (andlaus) [inline opm/material/fluidsystems/blackoilpvt/WetGasPvt.hpp]: yeah, some versions produce it, some don't. I guess that if you set `-Wextra`, exact float comparisons will always be complained about, though.

File: opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp (hunk 7 above)
Comment (andlaus) [inline opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp]: I mean the former, but I'm not sure because I consider having to eliminating such duplicates to be quite a hack. (i.e., IMO the input values can be considered to be broken if the x-values are not strictly monotonic in the context of these classes.)

Comment (babrodtk) [inline opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp]: I'm not entirely sure how this will work out with the splines, since these are end points and would hence affect more than the interval with zero slope. The issue is that these tables are created based upon evaluations of other tables again. 

I think the actual table is created in ```LiveOilPvt::updateSaturationPressure_(...)```, and if the table is monotonic (but not strictly increasing/decreasing), we may end up with "duplicate" Rs values (four in the case that I was examining) for different pSat values.

Comment (andlaus) [inline opm/material/fluidsystems/blackoilpvt/LiveOilPvt.hpp]: I think that it would be identical for splines, in particular for monotonic ones. that said, I agree that it is probably not a good idea.

**Comment by atgeirr** [comment]
Will merge now as it addresses a problem with model 2 (OPM/opm-simulators#1120). The missing part is that I wish it had unit tests. However, given the lack of unit tests here (other than API tests) it is nontrivial to add. This is a real problem. We have however allocated effort to address this within a reasonably short timeframe.
